src/communication/network_interface_sr.c

```diff
@@ -7268,6 +7269,368 @@ slogtb_dump_trantable (THREAD_ENTRY * thread_p, unsigned int rid, char *request,
+      css_send_reply_and_data_to_client (thread_p->conn_entry, rid, reply, OR_ALIGNED_BUF_SIZE (a_reply), buffer, size);
+    }
+  db_private_free_and_init (thread_p, buffer);
+}
+
+/*
+ * smmon_get_module_info -
+ *
+ * return:
+ *
+ *   rid(in):
+ *   request(in):
+ *   reqlen(in):
+ */
+void
+smmon_get_module_info (THREAD_ENTRY * thread_p, unsigned int rid, char *request, int reqlen)
+{
+  char *buffer, *ptr;
+  int size = 0;
+  OR_ALIGNED_BUF (OR_INT_SIZE + OR_INT_SIZE) a_reply;
+  char *reply = OR_ALIGNED_BUF_START (a_reply);
+  int error = NO_ERROR;
+  MMON_MODULE_INFO *module_info;
+  int module_index, module_count;
+
+  ptr = or_unpack_int (request, &module_index);
+
+  if (module_index == 0)
+    {
+      module_info = (MMON_MODULE_INFO *) db_private_alloc (thread_p, sizeof (MMON_MODULE_INFO) * MMON_MODULE_LAST);
+      module_count = MMON_MODULE_LAST;
+    }
+  else
+    {
+      module_info = (MMON_MODULE_INFO *) db_private_alloc (thread_p, sizeof (MMON_MODULE_INFO));
+      module_count = 1;
+    }
+
+  if (module_info == NULL)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, sizeof (MMON_MODULE_INFO) * module_count);
+      error = ER_OUT_OF_VIRTUAL_MEMORY;
+    }
+  else
+    {
+      error = mmon_aggregate_module_info (module_index, module_info);
+      if (error == NO_ERROR)
+	{
+	  /* before send information 
+	   * 1) check size
+	   * 2) allocate buffer
+	   * 3) packing information 
+	   * 4) deallocate MMON_MODULE_INFO */
+
+	  /* 1) check size 
+	   *    - calculate buffer size of packed values
+	   *    - module info -> component info w/ subcomponent info */
+	  for (int idx = 0; idx < module_count; idx++)
+	    {
+	      size += or_packed_string_length (module_info[idx].name, NULL);
+	      // Size of stat information (OR_INT64_SIZE * 3 + OR_INT_SIZE)
+	      // and size of num_comp (OR_INT_SIZE)
+	      size += OR_INT64_SIZE * 3 + OR_INT_SIZE * 2;
+
+              // *INDENT-OFF*
+              for (const auto &comp_info : module_info[idx].comp_info)
+                {
+                  size += or_packed_string_length (comp_info.name, NULL);
+                  size += OR_INT64_SIZE * 3 + OR_INT_SIZE * 2;
+
+                  for (const auto &subcomp_info : comp_info.subcomp_info)
+                    {
+                      size += or_packed_string_length (subcomp_info.name, NULL);
+                      size += OR_INT64_SIZE;
+                    }
+                }
+              // *INDENT-ON*
+	    }
+
+	  /* 2) allocate buffer */
+	  buffer = (char *) db_private_alloc (thread_p, size);
+
+	  if (buffer == NULL)
+	    {
+	      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, size);
+	      error = ER_OUT_OF_VIRTUAL_MEMORY;
+	    }
+	  else
+	    {
+	      /* 3) packing information
+	       *    - packing information with sequence of
+	       *      module info -> component info w/ subcomponent info */
+	      ptr = buffer;
+
```

Comment:
I guess that if you want to handle the module efficiently in the client-side, then you better to pack the number of the module like num_comp and num_subcomp.

Reply:
When the client(utility) recieves this payload, how to un-pack this since it doen't know the number of the modules? Please elaborate more detail.

Reply:
The client can know the total number of existing modules on the server through `MMON_MODULE_ID` in memory_monitor_common.h. Therefore, in the case of `mmon_get_module_info()`, depending on the `module_index`, it can request `resize()` either the entire set or just one module. Similarly, in the case of `mmon_get_module_info_summary()`, it performs resizing according to the `module_count` parameter passed from memmon function(std::min(5, MMON_MODULE_LAST) or MMON_MODULE_LAST). This size is the value the server will send as a result and remains constant, receive buffer must contains a fixed amount of information in the buffer or an error_code.